### PR TITLE
feat: check if hash is suspicious and put it in separate collection i…

### PR DESCRIPTION
**Pull request by pelavarthi**
…f it is

# Pull request

**Resolves:** #<<67>>

### CODE REVIEW

File: darkmoon/api/v1/metadata/views.py [hunk 1]
```diff
@@ -355,6 +362,51 @@ async def hash_comparison(
...
+                metadata.hashes.sha256,
+                metadata.hashes.sha512,
+            ]
+            inputHashes = [md5Hash, sha1Hash, sha256Hash, sha512Hash]
+            if dbHashes != inputHashes:
+                await susCollection.insert_one(obj.dict())
+                response.status_code = status.HTTP_406_NOT_ACCEPTABLE
+                return UploadListMetadataEntityResponse(
+                    message="The hash of your file did not match the hash of the \
+existing, file in the database. File placed in suspicious collection.",
```
Comment (Synthint) [inline darkmoon/api/v1/metadata/views.py:407]: This should probably be formatted a little cleaner to avoid the indentation change.

**Comment by Synthint** [comment]
As is there is no way to get anything from the suspicious hash collection, there should be a get endpoint to look through this collection.

This endpoint fails schema testing, that may have been missed before but when I ran it I got a 400 response code that is not defined in the schema. 

consider adding a test to ensure suspicious file metadata is being handled correctly

**Comment by Synthint** [comment]
The new test only checks if the correct status code is given and does not check if the item is correctly added to the new suspicious collection. 

A suspicious hash is put into the populated DB fixture but retrieval is not tested.

**Comment by asmitakarandikar** [comment]
Looks good to me

File: darkmoon/api/v1/metadata/views.py [hunk 2]
```diff
@@ -355,6 +403,47 @@ async def hash_comparison(
         sha256Hash = h_sha256.hexdigest()
         sha512Hash = h_sha512.hexdigest()
 
+        obj = MetadataEntity(
```
Comment (Synthint) [inline darkmoon/api/v1/metadata/views.py:406]: use the get_metadata() method from utils.py to gather and fill in all the information for the input file rather than leaving these sections blank.

Comment (pelavarthi) [reply darkmoon/api/v1/metadata/views.py:406]: We are working on this, but we are going to do it as part of the ticket to convert the hashComparison endpoint to use utils.py instead of as part of this ticket.

File: tests/integration/api/conftest.py [hunk 3]
```diff
@@ -87,17 +87,59 @@ def test_metadata_entity() -> (
     return file
 
 
+@pytest.fixture
+def test_suspicious_metadata_entity() -> (
+    dict[str, list[str] | dict[str, str] | dict[str, str | dict[str, str]]]
```
Comment (tclasen) [inline tests/integration/api/conftest.py:92]: Do you really need this complicated of a type? Would this be better off as a `dict[str, Any]` as traditional JSON is usually represented? Or maybe don't do `.dict()` and just return the model as is?

**Review by tclasen — CHANGES_REQUESTED** [review]

File: darkmoon/api/v1/metadata/views.py [hunk 4]
```diff
@@ -450,33 +458,33 @@ async def hash_comparison(
...
         sha512Hash = h_sha512.hexdigest()
 
-        obj = MetadataEntity(
+        obj = DocMetadataEntity(
             _id=PydanticObjectId(),
             name=[inputFileName],
             file_type=[inputFileType],
-            operating_system=[],
-            source_iso_name=[],
+            operating_system=[""],
```
Comment (tclasen) [inline darkmoon/api/v1/metadata/views.py:465]: An empty string isn't a valid operating system or a valid source iso name.

**Review by tclasen — CHANGES_REQUESTED** [review]

File: darkmoon/api/v1/metadata/views.py [hunk 5]
```diff
@@ -444,34 +447,16 @@ async def hash_comparison(
...
-            file_extension=[""],
-            hashes=Hashes(
-                md5=md5Hash,
-                sha1=sha1Hash,
-                sha256=sha256Hash,
-                sha512=sha512Hash,
-            ),
-        )
+
+        tmp_path = Path("tmpfile")
```
Comment (tclasen) [inline darkmoon/api/v1/metadata/views.py:451]: Please don't create your own implementation of the tempfile library. Please use the standard library version.

File: darkmoon/api/v1/metadata/views.py [hunk 6]
```diff
@@ -493,19 +478,26 @@ async def hash_comparison(
...
+                    data=data,
                 )
         search_query = {
             "$or": [
                 {
                     "name": inputFileName,
                 },
                 {
-                    "file_type": inputFileType,
+                    "type": inputFileType,
```
Comment (tclasen) [inline darkmoon/api/v1/metadata/views.py:500]: What is the value of searching for a name *OR* a filetype? Wouldn't the filetype have a ton of matches and cause false-positives in the suspicious files?

File: darkmoon/common/utils.py [hunk 7]
```diff
@@ -87,25 +87,22 @@ def get_all_exe_metadata(file: Path) -> dict[str, Any]:
     return exe_metadata
 
 
-def get_metadata(file: Path, source_iso: Path) -> dict[str, Any]:
+def get_metadata(file: Path, source_iso: str) -> dict[str, Any]:
```
Comment (tclasen) [inline darkmoon/common/utils.py:90]: Can you convert this function to return a Pydantic model? And on the return statement have it do something like `Model.parse_obj(data_fields)`. That way the data leaving this function is strongly typed.

File: tmpfile
Comment (tclasen) [inline tmpfile:1]: Please remove this file.

File: tests/integration/api/test_api.py [hunk 8]
```diff
@@ -130,7 +130,12 @@ def test_post_hash_comparison_failure(
     with TestClient(populated_app) as app:
         response = app.post(
             "/metadata/hashComparison",
-            files={"fileInput": open(test_hash_comparison_without_file, "rb")},
+            files={
+                "fileInput": open(test_hash_comparison_without_file, "rb"),
```
Comment (tclasen) [inline tests/integration/api/test_api.py:134]: This doesn't work the way you think it works. This is a context manager that has been left with a dangling reference. Please either read this into a variable and close the file, or use the context manager protocol to pass in the opened file handle.

File: pyproject.toml [hunk 9]
```diff
@@ -19,12 +19,13 @@ httpx = "^0.24.1"
 pydantic = "1.10.11"
 pefile = "^2023.2.7"
 uvicorn = { extras = ["standard"], version = "^0.23.1" }
-python-magic = "^0.4.27"
 beanie = "^1.19.2"
 typer = { extras = ["all"], version = "^0.9.0" }
 jinja2 = "^3.1.2"
 pytest = "^7.4.0"
 python-multipart = "^0.0.6"
+file-magic = "^0.4.1"
```
Comment (tclasen) [inline pyproject.toml:27]: Why did you add this import without using it? Technically it has a namespace collision with our other dependency of `python-magic`. Please choose one or the other.

**Review by tclasen — DISMISSED** [review]

File: tests/integration/api/test_api.py [hunk 10]
```diff
@@ -127,9 +127,34 @@ def test_post_hash_comparison_failure(
...
+                files={
+                    "fileInput": testFile,
+                },
+                params={
+                    "sourceIsoName": "Windows",
+                },
+            )
+            assert response.status_code == 404
+
+
```
Comment (Viraj-Ganguli) [inline tests/integration/api/test_api.py:143]: I don't see any 200 status codes. Are you guys asserting positive responses as well?

File: tests/integration/api/conftest.py [hunk 11]
```diff
@@ -93,17 +89,36 @@ def test_metadata_entity() -> (
     return file
 
 
+@pytest.fixture
+def test_suspicious_metadata_entity() -> dict[str, Any]:
+    """Represent a test metadata object."""
+    file: dict[str, Any] = DocMetadataEntity(
```
Comment (Viraj-Ganguli) [inline tests/integration/api/conftest.py:95]: Are you guys only testing doc metadata right now? If so that's not an issue as the schema changes should make it so you don't have to declare the type and can test whatever you want.

File: tests/integration/api/test_api.py (hunk 10 above)
Comment (XxXgamerguyXxX) [reply tests/integration/api/test_api.py:143]: Tory said that there was no point testing for the 200 status code since it doesn't give us any additional information.

**Review by DoodlePooge — APPROVED** [review]
